keyshare.go

```diff
@@ -1,29 +1,145 @@
 	"github.com/privacybydesign/gabi/internal/common"
 )
 
+var bigOne = big.NewInt(1)
+
+type (
+	// KeyshareCommitmentRequest contains the data the user must send to the keyshare server when it
+	// requests the keyshare server's contributions to the commitments, in the joint computation of
+	// the zero knowledge proof.
+	KeyshareCommitmentRequest struct {
+		HashedUserCommitments []byte `json:"hashedComms"`
+	}
+
+	// KeyshareResponseRequest contains the data the user must send to the keyshare server when it
+	// requests the keyshare server's contributions to the responses, in the joint computation of
+	// the zero knowledge proof.
+	KeyshareResponseRequest[T any] struct {
+		Context            *big.Int `json:"context,omitempty"`
+		Nonce              *big.Int `json:"nonce"`
+		UserResponse       *big.Int `json:"resp"`
+		IsSignatureSession bool     `json:"sig"`
+
+		// ChallengeInput contains the arguments used by the user to compute the
+		// HashedUserCommitments sent earlier in the commitment request.
+		ChallengeInput []KeyshareChallengeInput[T]
+	}
+
+	// KeyshareChallengeInput contains the user's contributions to the challenge, in the joint
+	// computation of the zero knowledge proof.
+	KeyshareChallengeInput[T any] struct {
+		// KeyID identifies the public key for this value and commitment. If nil, the keyshare
+		// server does not participate for this value and commitment.
+		KeyID *T `json:"key,omitempty"`
+
+		Value      *big.Int `json:"val"`
+		Commitment *big.Int `json:"comm"`
+
+		// OtherCommitments contain commitments for non-revocation proofs and range proofs
+		// (if present).
+		OtherCommitments []*big.Int `json:"otherComms,omitempty"`
+	}
+)
+
+// KeyshareUserCommitmentsHash computes the value h_W; that is, the commitment of the user to
+// its contributions to the challenge, in the joint computation of the zero knowledge proof.
+func KeyshareUserCommitmentsHash[T any](i []KeyshareChallengeInput[T]) ([]byte, error) {
+	bts, err := cbor.Marshal(i, cbor.EncOptions{})
+	if err != nil {
+		return nil, err
+	}
+	h := sha256.Sum256(bts)
+	return h[:], nil
+}
+
+// KeyshareResponse generates the keyshare response, using the keyshare secret and the user's
+// input in the keyshare protocol so far.
+func KeyshareResponse[T comparable](
+	secret *big.Int,
+	randomizer *big.Int,
+	req KeyshareCommitmentRequest,
+	res KeyshareResponseRequest[T],
+	keys map[T]*gabikeys.PublicKey,
+) (*ProofP, error) {
+	// Sanity checks
+	for i, k := range res.ChallengeInput {
+		if k.KeyID != nil && keys[*k.KeyID] == nil {
+			return nil, errors.Errorf("missing public key for element %d of challenge input", i)
+		}
+	}
+	if res.Context == nil {
+		res.Context = bigOne
+	}
+
+	hashContribs := make([]KeyshareChallengeInput[T], 0, len(res.ChallengeInput))
+	challengeContribs := make([]*big.Int, 0, len(res.ChallengeInput)*2)
+
+	// Assemble the input for the computation of h_W
+	for _, data := range res.ChallengeInput {
```

Comment:
What do you mean with "assembly"? The keyshare server (whose code this is) will receive the input parameters to `h_W = H(...)` from the irmaclient, so when when it gets them it can just forward it to this function. Or, do you mean the assembly of these parameters in the irmaclient code?

Reply:
Yep, I was thinking of generating/assembling `h_w` in the irmaclient code in irmago.

Reply:
My underlying point here is that I think this should be the same code since `h_w` is the same. That's why I thought it would be nice to have both the client and the server make use of the same function. Both `h_w` assemblies are now in `keyshare.go`. However, the code is still split and not very obvious the same. I guess you did not put the whole hashing functionality in one function because of linked surrounding code (e.g. `challengeContribs` needs the same for loops as the hashing etc).

It's still hard to read for me at least.
In `KeyshareResponse` the hash is introduced as `hashContribs := make([]KeyshareUserChallengeInput[T], 0, len(res.UserChallengeInput))` whereas in `KeyshareUserCommitmentRequest` it's `var hashInput []KeyshareUserChallengeInput[T]`. It would be nice if at the naming would be consistent.

Reply:
It's not the same code, because it can't be. The client computes the input parameters of `h_W` by invoking `Commit()` on its builders, which the keyshare server cannot do because it doesn't (and shouldn't) have those builders. Instead, the keyshare server directly takes whatever the client claims were its input parameters to `h_W = H(...)`, as sent by the client when it requests the proof response from the keyshare server.

So the client and keyshare server can't use the same code to compute `h_W` because of their different roles and knowledge. This is not new in `gabi`, a similar thing happens with the commitment in creating or verifying a zero-knowledge proof:
* The client computes the commitment as (schematically) `commitment := base^randomizer`;
* When verifying the proof, the verifier recomputes the commitment without having access to the `randomizer` value (which is private to the client), but instead from the proof that it receives from the client.

> In `KeyshareResponse` the hash is introduced as `hashContribs := make(...)` whereas in `KeyshareUserCommitmentRequest` it's `var hashInput []KeyshareUserChallengeInput[T]`. It would be nice if at the naming would be consistent.

I now use the `make(...)` syntax in `KeyshareUserCommitmentRequest` as well (apart from consistency that also avoids potential slice growing by creating the slice with the appropriate capacity). As to naming, per your other comment I have removed the `hashInput` variable from `KeyshareResponse` because it's not necessary there, so that inconsistency is now gone as well 😅